### PR TITLE
Cleanup microcluster/k8s when scaling down cluster nodes

### PR DESCRIPTION
**Pull request by bschimke95**
## Summary

Remove nodes from microcluster and k8s when scaling down the cluster.

## Notable changes

* Moved k8sd-proxy requests into generic separate function
* Added k8sd-proxy blacklist options to ensure the proxy is not targeting the machine that is removed
* Added `RemoveMachineFromCluster` method to workload cluster

## Tests

Manual tests for now:
* Created 4 CP node cluster - removed 1 of them again
  -> verified on remaining nodes that node is removed from microcluster and roles are updated
* Created 1 CP, 1 worker cluster - removed worker
  -> verified that worker is removed from k8s

### CODE REVIEW

File: bootstrap/controllers/ck8sconfig_controller.go [hunk 1]
```diff
@@ -235,7 +235,7 @@ func (r *CK8sConfigReconciler) joinControlplane(ctx context.Context, scope *Scop
 		return fmt.Errorf("failed to create remote cluster client: %w", err)
 	}
 
-	joinToken, err := workloadCluster.NewControlPlaneJoinToken(ctx, *authToken, scope.Config.Spec.ControlPlaneConfig.MicroclusterPort, scope.Config.Name)
+	joinToken, err := workloadCluster.NewControlPlaneJoinToken(ctx, *authToken, scope.Config.Spec.ControlPlaneConfig.MicroclusterPort, machine.Name)
```
Comment (bschimke95) [inline bootstrap/controllers/ck8sconfig_controller.go:238]: drive-by fix: the worker hostname and token name did not match

File: controlplane/controllers/scale.go [hunk 2]
```diff
@@ -159,6 +160,29 @@ func (r *CK8sControlPlaneReconciler) scaleDownControlPlane(
...
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to lookup auth token: %w", err)
+	}
+
+	if authToken == nil {
+		return ctrl.Result{}, fmt.Errorf("auth token not yet generated")
+	}
+
+	// TODO: Get from CK8sConfig
+	microclusterPort := 2380
```
Comment (bschimke95) [inline controlplane/controllers/scale.go:180]: I could not really figure out what is the best way to access the cluster config from this controller.

File: controlplane/main.go [hunk 3]
```diff
@@ -66,7 +66,7 @@ func main() {
 	flag.DurationVar(&syncPeriod, "sync-period", 10*time.Minute,
 		"The minimum interval at which watched resources are reconciled (e.g. 15m)")
 
-	flag.DurationVar(&k8sdDialTimeout, "k8sd-dial-timeout-duration", 10*time.Second,
+	flag.DurationVar(&k8sdDialTimeout, "k8sd-dial-timeout-duration", 60*time.Second,
```
Comment (bschimke95) [inline controlplane/main.go:69]: Currently, the node-removal k8s-snap endpoint waits until the node is removed. This may take more than 10s

File: bootstrap/controllers/ck8sconfig_controller.go [hunk 4]
```diff
@@ -308,7 +308,7 @@ func (r *CK8sConfigReconciler) joinWorker(ctx context.Context, scope *Scope) err
 		return fmt.Errorf("failed to create remote cluster client: %w", err)
 	}
 
-	joinToken, err := workloadCluster.NewWorkerJoinToken(ctx, *authToken, scope.Config.Spec.ControlPlaneConfig.MicroclusterPort, scope.Config.Name)
+	joinToken, err := workloadCluster.NewWorkerJoinToken(ctx, *authToken, scope.Config.Spec.ControlPlaneConfig.MicroclusterPort, machine.Name)
```
Comment (neoaggelos) [inline bootstrap/controllers/ck8sconfig_controller.go:311]: TLDR: for worker nodes, please use an empty string instead `""`, such that any hostname is accepted. 

This is very important for worker nodes, as different infrastructures will have machines where hostname and machine name do not match by design.

for example, in AWS, machine names will be `$cluster-control-plane-$foobar`, and then hostnames will be `ip-$1-$2-$3-$4`. this is something we _need_, since the external cloud providers sometimes necessitate that the node matches specific criteria, see https://cloud-provider-aws.sigs.k8s.io/prerequisites/#node-name-conventions

File: controlplane/controllers/scale.go (hunk 2 above)
Comment (neoaggelos) [reply controlplane/controllers/scale.go:180]: NOTE: microcluster port is not something that changes, and therefore should be handled in the workload cluster, not in every method.

File: controlplane/controllers/machine_controller.go [hunk 5]
```diff
@@ -36,15 +34,15 @@ func (r *MachineReconciler) SetupWithManager(ctx context.Context, mgr ctrl.Manag
...
 
 	// NOTE(neoaggelos): See note below
-	/**
 	if r.managementCluster == nil {
 		r.managementCluster = &ck8s.Management{
-			Client:          r.Client,
-			EtcdDialTimeout: r.EtcdDialTimeout,
-			EtcdCallTimeout: r.EtcdCallTimeout,
+			Client: r.Client,
+			/*
```
Comment (neoaggelos) [inline controlplane/controllers/machine_controller.go:40]: let's set the dial timeout here

File: controlplane/controllers/scale.go [hunk 6]
```diff
@@ -159,6 +160,27 @@ func (r *CK8sControlPlaneReconciler) scaleDownControlPlane(
...
+	}
+
+	authToken, err := token.Lookup(ctx, r.Client, clusterObjectKey)
+	if err != nil {
+		return ctrl.Result{}, fmt.Errorf("failed to lookup auth token: %w", err)
+	}
+
+	if authToken == nil {
+		return ctrl.Result{}, fmt.Errorf("auth token not yet generated")
+	}
```
Comment (neoaggelos) [inline controlplane/controllers/scale.go:177]: The token lookup feels a bit out of place here, maybe it could be included in the GetWorkloadCluster, or perhaps the `RemoveMachineFromCluster` call. But that's ok, let's sleep on this and see how the code evolves before any changes.

File: controlplane/controllers/scale.go [hunk 7]
```diff
@@ -159,6 +160,27 @@ func (r *CK8sControlPlaneReconciler) scaleDownControlPlane(
...
+
+	err = workloadCluster.RemoveMachineFromCluster(ctx, machineToDelete, *authToken)
+	if err != nil {
```
Comment (neoaggelos) [inline controlplane/controllers/scale.go:180]: otherwise err trickles down to the rest of the code, let's avoid this

```suggestion
	if err := workloadCluster.RemoveMachineFromCluster(ctx, machineToDelete, *authToken); err != nil {
```

File: pkg/ck8s/workload_cluster.go [hunk 8]
```diff
@@ -172,13 +175,24 @@ func getNodeInternalIP(node *corev1.Node) (string, error) {
 	return "", fmt.Errorf("unable to find internal IP for node %s", node.Name)
 }
 
-func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context) (*K8sdClient, error) {
+type k8sdProxyOptions struct {
+	// BlacklistedControlPlanes is a list of control plane nodes which k8sd proxy should not use.
+	// This is useful when a control plane node is being removed and we want to avoid using it
+	BlacklistedControlPlanes []string
```
Comment (neoaggelos) [inline pkg/ck8s/workload_cluster.go:181]: i'd make this a `IgnoreNodes map[string]struct{}`

File: pkg/ck8s/workload_cluster.go [hunk 9]
```diff
@@ -172,13 +175,24 @@ func getNodeInternalIP(node *corev1.Node) (string, error) {
...
+
+// GetK8sdProxyForControlPlane returns a k8sd proxy client for the control plane.
+func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context, options k8sdProxyOptions) (*K8sdClient, error) {
 	cplaneNodes, err := w.getControlPlaneNodes(ctx)
 	if err != nil {
 		return nil, fmt.Errorf("failed to get control plane nodes: %w", err)
 	}
 
 	for _, node := range cplaneNodes.Items {
+		if slices.Contains(options.BlacklistedControlPlanes, node.Name) {
```
Comment (neoaggelos) [inline pkg/ck8s/workload_cluster.go:192]: then this a `_, ok := opts.IgnoredNodes[node.Name]; ok`

File: pkg/ck8s/workload_cluster.go [hunk 10]
```diff
@@ -192,58 +206,93 @@ func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context) (*K8sdClient
...
-		Error    string                     `json:"error"`
-		Metadata apiv1.GetJoinTokenResponse `json:"metadata"`
+		Error    string          `json:"error"`
+		Metadata json.RawMessage `json:"metadata"`
 	}
 
-	requestBody, err := json.Marshal(apiv1.GetJoinTokenRequest{Name: name, Worker: worker})
+	// FIXME: Get this from the workload cluster's configuration instead of hardcoding it.
+	// See https://github.com/canonical/cluster-api-k8s/pull/8#discussion_r1645536362 for context.
+	microclusterPort := 2380
```
Comment (neoaggelos) [inline pkg/ck8s/workload_cluster.go:250]: please, let's take this in when calling `GetWorkloadCluster()`, do not hardcode this.

File: pkg/ck8s/workload_cluster.go [hunk 11]
```diff
@@ -192,58 +206,93 @@ func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context) (*K8sdClient
...
+
+	if response == nil {
+		// Nothing to decode
+		return nil
+	}
 
-	var wrappedResp wrappedResponse
-	if err := json.NewDecoder(res.Body).Decode(&wrappedResp); err != nil {
-		return "", fmt.Errorf("failed to parse HTTP response: %w", err)
+	respBody := wrappedResponse{}
```
Comment (neoaggelos) [inline pkg/ck8s/workload_cluster.go:284]: more idiomatic
```suggestion
	var respBody wrappedResponse
```

Comment (neoaggelos) [reply pkg/ck8s/workload_cluster.go:284]: also, since we have `requestBody` earlier, let this be `responseBody`

File: controlplane/controllers/scale.go (hunk 2 above)
Comment (bschimke95) [reply controlplane/controllers/scale.go]: added it to the `GetWorkloadCluster` call

File: controlplane/controllers/scale.go (hunk 6 above)
Comment (bschimke95) [reply controlplane/controllers/scale.go]: Moved it to the GetWorkloadCluster call, feels quite right there.

File: pkg/ck8s/workload_cluster.go [hunk 12]
```diff
@@ -43,8 +43,10 @@ type WorkloadCluster interface {
 	ClusterStatus(ctx context.Context) (ClusterStatus, error)
 	UpdateAgentConditions(ctx context.Context, controlPlane *ControlPlane)
 	UpdateEtcdConditions(ctx context.Context, controlPlane *ControlPlane)
-	NewControlPlaneJoinToken(ctx context.Context, authToken string, microclusterPort int, name string) (string, error)
-	NewWorkerJoinToken(ctx context.Context, authToken string, microclusterPort int, name string) (string, error)
+	NewControlPlaneJoinToken(ctx context.Context, name string) (string, error)
+	NewWorkerJoinToken(ctx context.Context, name string) (string, error)
```
Comment (neoaggelos) [inline pkg/ck8s/workload_cluster.go:47]: final note, let's drop the `name` from the interface here. We expect this to always be `""` so let's hardcode it into the implementation. then callers don't need to care.

_If we ever change this_, we'll add a V2, or change the interface, then callers must adapt and we don't miss any required updates in the code.

File: pkg/ck8s/workload_cluster.go [hunk 13]
```diff
@@ -67,10 +69,11 @@ type WorkloadCluster interface {
 // Workload defines operations on workload clusters.
 type Workload struct {
 	WorkloadCluster
-
+	AuthToken           string
 	Client              ctrlclient.Client
 	ClientRestConfig    *rest.Config
 	K8sdClientGenerator *k8sdClientGenerator
+	MicroclusterPort    int
```
Comment (neoaggelos) [inline pkg/ck8s/workload_cluster.go:76]: any objections to making `MicroclusterPort` and `AuthToken` private fields?

File: pkg/ck8s/workload_cluster.go [hunk 14]
```diff
@@ -192,58 +206,91 @@ func (w *Workload) GetK8sdProxyForControlPlane(ctx context.Context) (*K8sdClient
...
-	if res.StatusCode != 200 {
-		return "", fmt.Errorf("request for join token request failed: %s", wrappedResp.Error)
+
+	var responseBody wrappedResponse
+	if err := json.NewDecoder(res.Body).Decode(&responseBody); err != nil {
+		return fmt.Errorf("failed to parse HTTP response: %w", err)
+	}
+	if res.StatusCode != http.StatusOK {
+		return fmt.Errorf("HTTP request failed with status code: %d (%s)", res.StatusCode, responseBody.Error)
+	}
```
Comment (neoaggelos) [inline pkg/ck8s/workload_cluster.go:285]: should we not first check the status coe, then start decoding things (or skip decoding if the response is `nil`)?

**Review by neoaggelos — APPROVED** [review]
LGTM!

File: pkg/ck8s/management_cluster.go [hunk 15]
```diff
@@ -87,10 +88,25 @@ func (m *Management) GetWorkloadCluster(ctx context.Context, clusterKey client.O
...
+		return nil, fmt.Errorf("failed to lookup auth token: %w", err)
+	}
+
+	if authToken == nil {
+		return nil, fmt.Errorf("auth token not yet generated")
+	}
+
+	if microclusterPort == 0 {
+		microclusterPort = 2380
+	}
```
Comment (neoaggelos) [inline pkg/ck8s/management_cluster.go:102]: the more i see this, the more i think we should approach this differently.

Make the MicroclusterPort field a `*int`, then add a `GetMicroclusterPort()` method, which returns the field (if set), otherwise the default value (in this case 2380)`

then use `.GetMicroclusterPort()` for all invocations and remove this.

Comment (bschimke95) [reply pkg/ck8s/management_cluster.go]: done
